dask/dataframe/operation.py

```diff
@@ -0,0 +1,596 @@
+            arg.name: arg for arg in self.args if isinstance(arg, CollectionOperation)
+        }
+        divisions = divisions or (None,) * (
+            len(next(iter(self._dependencies.values()))) + 1
+        )
+        self._divisions = tuple(divisions)
+        self._columns = columns
+        self.kwargs = kwargs
+
+    def copy(self):
+        return type(self)(
+            self.func,
+            self.meta,
+            self.args,
+            columns=self.columns,
+            divisions=self.divisions,
+            label=self.label,
+            **self.kwargs,
+        )
+
+    def subgraph(self, keys: list[tuple]) -> tuple[dict, dict]:
+        # Start by populating the key dependencies.
+        # If the dependency is "fusable", we add it to a
+        # seperate `fusable_graph` dictionary.
+        fusable_graph = {}
+        dep_keys = {}
+        for name, dep in self.dependencies.items():
+            if isinstance(dep, (DataFrameMapOperation, DataFrameCreation)):
+                # These dependencies are "fusable"
+                _graph, _keys = dep.subgraph([(name, key[1]) for key in keys])
+                fusable_graph.update(_graph)
+                dep_keys.update(_keys)
+            else:
+                # Not fusing these dependencies
+                dep_keys[dep] = [(name, key[1]) for key in keys]
+
+        # Now we just need to update the graph with the
+        # current DataFrameMapOperation tasks. If any dependency
+        # keys are in `fusable_graph`, we use the corresponding
+        # element from `fusable_graph` (rather than the task key).
+        dsk: dict[tuple, tuple] = {}
+        for key in keys:
+            name, index = key
+            assert name == self.name
+
+            task = [self.func]
+            for arg in self.args:
+                if isinstance(arg, CollectionOperation):
+                    dep_key = (arg.name, index)
+                    task.append(fusable_graph.pop(dep_key, dep_key))
+                else:
+                    task.append(arg)
+
+            if self.kwargs:
+                dsk[key] = (
+                    apply,
+                    task[0],
+                    task[1:],
+                    self.kwargs,
+                )
+            else:
+                dsk[key] = tuple(task)
+
+        dsk.update(fusable_graph)
+        return dsk, dep_keys
+
+    @property
+    def dependencies(self) -> Mapping[str, CollectionOperation]:
+        return self._dependencies
+
+    def regenerate(self, new_dependencies: dict, **new_kwargs):
+        kwargs = {
+            "divisions": self.divisions,
+            "label": self.label,
+            "columns": self.columns,
+            **self.kwargs,
+        }
+        kwargs.update(new_kwargs.get(self.name, {}))
+        args = [
+            new_dependencies[arg.name] if isinstance(arg, CollectionOperation) else arg
+            for arg in self.args
+        ]
+        return type(self)(
```

Comment:
I think this needs some careful design thought, but I don't know enough about the different objects that one will need. I think there are two ways this can go:
1. No subclasses of the base "operation" class have their own init method: so now it must be that the base class can regenerate.
2. Subclasses have their own init, but conforming to a contract that the base class advertises.

For option two, arguments to init would separate into "child" data (that is instances of `Operation`), and "non-child" (other parameters). So one would either need a contract for how to pass this, or a way of advertising which slots are which.